src/reflect/scala/reflect/internal/TreeInfo.scala

```diff
@@ -224,7 +224,7 @@ abstract class TreeInfo {
       }
       def isWarnableSymbol = {
         val sym = tree.symbol
-        (sym == null) || !(sym.isModule || sym.isLazy || definitions.isByNameParamType(sym.tpe_*)) || {
+        (sym == null) || !(sym.isSynthetic  || sym.decodedName.startsWith("$") || sym.isModule || sym.isLazy || definitions.isByNameParamType(sym.tpe_*)) || {
```

Comment:
Are both additions necessary for `$fullAccessPath`?

Reply:
For this particular one only the hackier one `sym.decodedName.startsWith("$")` is needed. I'm trying to exempt all synthetic symbols though.

Reply:
We should do something more fine-grained than that.  It's `$line` I think?  There's already 

```scala
    val INTERPRETER_IMPORT_WRAPPER     = "$iw"
    val INTERPRETER_WRAPPER            = "$read"
```

perhaps we just need to add "$line" to that?  There's also `isLineReadVal`, which it looks like needs some attention from being merged from the 2.12 repl to the 2.13 repl.

Reply:
REPL is not the only code that generates synthetic code with dollar sign tho. I'd rather fix "pure expression in any synthetic looking" code problem than this narrow REPL situation.

Reply:
That would be good, but `sym.decodedName.startsWith("$")` is an overreach IMO.

Reply:
How so? "pure expression does nothing" is a linting warning at best, and by convention `$` variable names are not used by normal human-written programs.

Reply:
Is every tree symbol a variable name?  Does every tree symbol with a `$` make it not human-facing?

I've been looking at parts of the exhaustivity checker and there's a 1001 ways where the check can bail out and give no heads-up.  So I'm keen that we don't reach for the biggest tool in the toolbox here.

To be fair the problem is the REPL should take care of communicating its synthetic code, rather than having to reverse-engineer it here.  Have you seen `wrapperCleanup`?  Could we do something like that to address the original issue?

Reply:
> Is every tree symbol a variable name? Does every tree symbol with a `$` make it not human-facing?

I'd claim that for the purpose of warning (this method is nested in `isPureExprForWarningPurposes`), sure it's ok to ignore `$`.

It would be ideal if the symbols created by the REPLs are automatically flagged as synthetic. I guess from my point of view, I sent in quick fixes for the bug that was reported. It's a lint, and likely not going to break anyone's code. I'd rather not spend further time on this, so feel free to send in another PR I guess.